### ISSUE TITLE
ObjectDisposedException on requests during shell restart

### ISSUE THREAD

**Issue by orchardbot** [OrchardCMS/Orchard#4569]
@Piedone created:
https://orchard.codeplex.com/workitem/20740

When you e.g. enable/disable a feature the shell will be restarted. If during the restart you try to access any page of the site you'll get an ObjectDisposedException from various places because the shell was already disposed. The exception says "Instances cannot be resolved and nested lifetimes cannot be created from this LifetimeScope as it has already been disposed."

This is caused by DefaultOrchardHost.ActivateShell() from ln 285: first the current shell is disposed, then the new shell activated. Any request or background task between the two will cause the exception.

I just tested it with latest 1.8.x (vanilla instance) and although you should be fast (and probably lucky) to experience it you can repro it by enabling/disabling a feature and immediately refreshing another page.

A possible fix is to to only dispose the shell after the new one is created, see:

                var context = _shellContextFactory.CreateShellContext(settings);

                // dispose previous context
                shellContext.Shell.Terminate();
                shellContext.LifetimeScope.Dispose();

                // activate and register modified context
                _shellContexts = _shellContexts.Where(shell => shell.Settings.Name != settings.Name).Union(new[] { context });

                context.Shell.Activate();

                _runningShellTable.Update(settings);
This is pretty safe as the shell is activated after the previous one is disposed (this mitigates any possible issue that could be cause by the same shell running as multiple instances - don't know if there can be any such issues). However since there is a small interval (<1s) between the old shell being disposed and the new one updates in RunningShellTable it can cause users to experience 404s on page views (since RunningShellTable won't find the shell in that moment).

Another possible fix:

                var context = _shellContextFactory.CreateShellContext(settings);

                context.Shell.Activate();

                // activate and register modified context
                _shellContexts = _shellContexts.Where(shell => shell.Settings.Name != settings.Name).Union(new[] { context });

                _runningShellTable.Update(settings);

                // dispose previous context
                shellContext.Shell.Terminate();
                shellContext.LifetimeScope.Dispose();
This is probably a bit more risky since the old and the new shell runs in parallel for a short time. Note that this activates the shell before adding it to _shellContexts (this is actually how ActivateShell() also does it). However this still causes 404s to be seen occasionally.

**Comment by Piedone** [comment]
Implemented a possible fix here: https://github.com/OrchardCMS/Orchard/tree/issue/20740 This will block any requests to a tenant as long as it is being restarted, see the delta: https://github.com/OrchardCMS/Orchard/compare/1.x...issue/20740

**Comment by kumards** [comment]
If this helps, I am getting this exception "System.ObjectDisposedException: Instances cannot be resolved and nested lifetimes cannot be created from this LifetimeScope as it has already been disposed." when my 
tasks gets executed by IProcessingEngine.ExecuteNextTask method.

**Comment by Piedone** [comment]
@kumards that looks something different (despite the exception being the same). It may have to do something you've implemented in your tasks.

**Comment by kumards** [comment]
@Piedone  after debugging I found the below line 
part.OwnerField.Loader(() => _contentManager.Get<IUser>(part.Record.OwnerId));
from CommonPartHandler.LazyLoadHandlers method gets ContentManager with disposed _context and hence the exception I stated in my previous comment. I do not know the root cause yet, I will let you know If I found.

**Comment by Piedone** [comment]
@kumards What happens if you change the _contentManager injection there to Work<IContentManager> and use it as _contentManagerWork.Value? My guess is that you are using content items retrieved in a different work context (but sharing content items between WCs is not suppproted). Any way, please open a discussion about this in the forums and link it here, I don't think this is related to this issue.

**Comment by kumards** [comment]
@Piedone  Good catch. **sharing content items between WCs is not suppproted** solved my problem.

Injecting ContentManager as Work<IContentManager> did not solve the problem, its value becomes null.

**Comment by Piedone** [comment]
We should probably have something to make this more apparent: documentation is not the key here as sharing objects between WCs is most of the time subtle and involuntary in my experience. Maybe we could wrap ObjectDisposedExceptions into another exception explaining the problem...

**Comment by Piedone** [comment]
@sebastienros this is not fixed AFAIK. Why did you close the issue? We can postpone it to later.

**Comment by Piedone** [comment]
Since https://github.com/OrchardCMS/Orchard/pull/6611 was merged we can close this.
